pkg/crc/api/api.go

```diff
@@ -0,0 +1,91 @@
+package api
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net"
+
+	"github.com/code-ready/crc/pkg/crc/logging"
+)
+
+type ArgsType map[string]string
+type handlerFunc func(ArgsType) string
+
+type CrcApiServer struct {
+	listener net.Listener
+	handlers map[string]handlerFunc // relates commands to handler func
+}
+
+// commandRequest struct is used to decode the json request from tray
+type commandRequest struct {
+	Command string            `json:"command"`
+	Args    map[string]string `json:"args,omitempty"`
+}
+
+func CreateApiServer(socketPath string) CrcApiServer {
+	lnr, err := net.Listen("unix", socketPath)
+	if err != nil {
+		logging.Error("Failed to create socket: ", err.Error())
+		return CrcApiServer{}
+	}
+	apiServer := CrcApiServer{
+		listener: lnr,
+		handlers: map[string]handlerFunc{
+			"start":         startHandler,
+			"stop":          stopHandler,
+			"status":        statusHandler,
+			"delete":        deleteHandler,
+			"version":       versionHandler,
+			"webconsoleurl": webconsoleURLHandler,
+		},
+	}
+	return apiServer
+}
+
+func (api CrcApiServer) Serve() {
+	for {
+		conn, err := api.listener.Accept()
+		if err != nil {
+			logging.Error("Error establishing communication: ", err.Error())
+		}
+		go api.handleConnection(conn)
+	}
+}
+
+func (api CrcApiServer) handleConnection(c net.Conn) {
+	inBuffer := make([]byte, 1024)
+	var outBuffer bytes.Buffer
+	var req commandRequest
+	numBytes, err := c.Read(inBuffer)
+	if err != nil || numBytes == 0 {
+		logging.Error("Error reading from socket")
+	}
+	logging.Debug("Received Request:", string(inBuffer[0:numBytes]))
+	err = json.Unmarshal(inBuffer[0:numBytes], &req)
+	if err != nil {
+		logging.Error("Error decoding request: ", err.Error())
+	}
+
+	if h, ok := api.handlers[req.Command]; ok {
```

Comment:
can this be simplified? a lot of this looks like duplication between having a command and an unknown command.

Reply:
Will try to simplify, although the duplication is because in both cases we write something to the socket, in case of a valid command the result, and an error msg otherwise.

Reply:
so extract a function in which you can supply a message based on the check?

Reply:
Yes, this is what i did.